drivers/virtio/virtio_mmio.c

```diff
@@ -0,0 +1,419 @@
+			    k_mem_phys_addr(virtqueue->desc) & UINT32_MAX);
+	virtio_mmio_write32(dev, VIRTIO_MMIO_QUEUE_DESC_HIGH,
+			    k_mem_phys_addr(virtqueue->desc) >> 32);
+	virtio_mmio_write32(dev, VIRTIO_MMIO_QUEUE_DRIVER_LOW,
+			    k_mem_phys_addr(virtqueue->avail) & UINT32_MAX);
+	virtio_mmio_write32(dev, VIRTIO_MMIO_QUEUE_DRIVER_HIGH,
+			    k_mem_phys_addr(virtqueue->avail) >> 32);
+	virtio_mmio_write32(dev, VIRTIO_MMIO_QUEUE_DEVICE_LOW,
+			    k_mem_phys_addr(virtqueue->used) & UINT32_MAX);
+	virtio_mmio_write32(dev, VIRTIO_MMIO_QUEUE_DEVICE_HIGH,
+			    k_mem_phys_addr(virtqueue->used) >> 32);
+
+	virtio_mmio_write32(dev, VIRTIO_MMIO_QUEUE_READY, 1);
+
+	return 0;
+}
+
+static int virtio_mmio_set_virtqueues(const struct device *dev, uint16_t queue_count,
+				      virtio_enumerate_queues cb, void *opaque)
+{
+	struct virtio_mmio_data *data = dev->data;
+
+	data->virtqueues = k_malloc(queue_count * sizeof(struct virtq));
+	if (!data->virtqueues) {
+		LOG_ERR("failed to allocate virtqueue array");
+		return -ENOMEM;
+	}
+	data->virtqueue_count = queue_count;
+
+	for (int i = 0; i < queue_count; i++) {
+		virtio_mmio_write32(dev, VIRTIO_MMIO_QUEUE_SEL, i);
+
+		const uint16_t queue_size =
+			cb(i, virtio_mmio_read32(dev, VIRTIO_MMIO_QUEUE_SIZE_MAX), opaque);
+
+		int ret = virtq_create(&data->virtqueues[i], queue_size);
+
+		if (ret != 0) {
+			for (int j = 0; j < i; i++) {
+				virtq_free(&data->virtqueues[j]);
+			}
+			return ret;
+		}
+		ret = virtio_mmio_set_virtqueue(dev, i, &data->virtqueues[i]);
+		if (ret != 0) {
+			for (int j = 0; j < i; i++) {
+				virtq_free(&data->virtqueues[j]);
+			}
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int virtio_mmio_init_virtqueues(const struct device *dev, uint16_t num_queues,
+				       virtio_enumerate_queues cb, void *opaque)
+{
+	return virtio_mmio_set_virtqueues(dev, num_queues, cb, opaque);
+}
+
+static void virtio_mmio_finalize_init(const struct device *dev)
+{
+	virtio_mmio_write_status_bit(dev, DEVICE_STATUS_DRIVER_OK);
+}
+
+static const struct virtio_driver_api virtio_mmio_driver_api = {
+	.get_virtqueue = virtio_mmio_get_virtqueue,
+	.notify_virtqueue = virtio_mmio_notify_queue,
+	.get_device_specific_config = virtio_mmio_get_device_specific_config,
+	.read_device_feature_bit = virtio_mmio_read_device_feature_bit,
+	.write_driver_feature_bit = virtio_mmio_write_driver_feature_bit_range_check,
+	.commit_feature_bits = virtio_mmio_commit_feature_bits,
+	.init_virtqueues = virtio_mmio_init_virtqueues,
+	.finalize_init = virtio_mmio_finalize_init,
+};
+
+static int virtio_mmio_init_common(const struct device *dev)
+{
+	DEVICE_MMIO_NAMED_MAP(dev, reg_base, K_MEM_CACHE_NONE);
+
+	const uint32_t magic = virtio_mmio_read32(dev, VIRTIO_MMIO_MAGIC_VALUE);
+
+	if (magic != VIRTIO_MMIO_MAGIC) {
+		LOG_ERR("Invalid magic value %x", magic);
+		return -EINVAL;
+	}
+
+	const uint32_t version = virtio_mmio_read32(dev, VIRTIO_MMIO_VERSION);
+
+	if (version < VIRTIO_MMIO_SUPPORTED_VERSION) {
```

Comment:
```suggestion
	if (version != VIRTIO_MMIO_SUPPORTED_VERSION) {
```

It would probably be better to have this as an exact check in case new revision is introduced by the standard.

Reply:
Is this field broken up into semantic version numbers? If so, maybe the major version should be the deciding field. Typically minor versions should still be still compatible.

Reply:
@cfriedt 

This value is defined as single 32 bit number.
So, it can't take semver format.
(Now only defined 2 versions. 0x1 for legacy, 0x2 for current spec.)

@fkokosinski 

Thank you for your pointing out.
Fixed it.